budgetportal/models.py

```diff
@@ -768,8 +769,324 @@ def get_expenditure_over_time(self):
+                                                                           'Direct charge against the National '
+                                                                           'Revenue Fund')
+            total_positive_virement_change = 0
+            for c in cells_for_virements:
+                value = c['value.sum']
+                if value > 0:
+                    total_positive_virement_change += value
+            virements = {
+                'label': 'Value of virements and shifts due to savings',
+                'amount': int(total_positive_virement_change),
+                'percentage': 100 * float(total_positive_virement_change) / float(total_voted)
+            }
+        return virements if virements else None
+
+    def _get_budget_special_appropriations(self, openspending_api, total_voted):
+        result_for_special_appropriations = openspending_api.aggregate(
+            cuts=[
+                openspending_api.get_financial_year_ref() + ':' + self.get_financial_year().get_starting_year(),
+                openspending_api.get_adjustment_kind_ref() + ':' + '"Special appropriation"',
+
+            ],
+            drilldowns=[
+                openspending_api.get_department_name_ref()
+            ])
+        result_for_special_appropriations = openspending_api.filter_dept(result_for_special_appropriations, self.name)
+
+        total_special_appropriations = 0
+        for cell in result_for_special_appropriations['cells']:
+            if cell['value.sum']:
+                total_special_appropriations += cell['value.sum']
+
+        if total_special_appropriations:
+            return {
+                'amount': total_special_appropriations,
+                'percentage': (float(total_special_appropriations) / float(total_voted)) * 100
+            }
+        else:
+            return None
+
+    def _get_budget_direct_charges(self, openspending_api):
+        result_for_direct_charges = openspending_api.aggregate(
+            cuts=[
+                openspending_api.get_financial_year_ref() + ':' + self.get_financial_year().get_starting_year(),
+                openspending_api.get_programme_name_ref() + ':' + '"Direct charge against the National Revenue Fund"',
+            ],
+            drilldowns=[
+                openspending_api.get_phase_ref(),
+                openspending_api.get_subprogramme_name_ref(),
+                openspending_api.get_department_name_ref(),
+                openspending_api.get_adjustment_kind_ref(),
+            ],
+            order=[openspending_api.get_subprogramme_name_ref()])
+        result_for_direct_charges = openspending_api.filter_dept(result_for_direct_charges, self.name)
+
+        subprog_ref = openspending_api.get_subprogramme_name_ref()
+        phase_ref = openspending_api.get_phase_ref()
+        kind_ref = openspending_api.get_adjustment_kind_ref()
+        subprog_dict = {}
```

Comment:
since we stick them into a dict we lose ordering - we should ensure consistent ordering - one option is to use OrderedDict here. Another is to sort later.

Reply:
thanks, did not consider ordering. i've implemented an ordereddict here as suggested